### ISSUE TITLE
Changing line thickness of dashed lines gives very different results on FF and Chrome, with the Chome one being near useless

### ISSUE THREAD

**Issue by roel-v** [Pattern-Projector/pattern-projector#195]
Since pictures say more than a thousand words:

![2024-04-07 14 53 23](https://github.com/Pattern-Projector/pattern-projector/assets/568567/ec62a04e-a6cf-4095-b016-f0122d57bf5d)
![2024-04-07 14 53 32](https://github.com/Pattern-Projector/pattern-projector/assets/568567/452561a8-359a-4577-8f1a-fa389ad9bc38)

Top one is on Firefox, bottom on Chrome. The line in the original pattern is a very finely dashed line, and it seems that whatever is used to increase the line thickness, the rendering engine tries to approximate capturing that nuance by incorporating it in the antialiasing. On Chrome this causes lines to drop out completely. If I move the pattern ever so slightly, the lines pop back into display, but then others become much less visible and so on.

I think what you want here is to not have lines drawn dashed at all, just continuous. I'm not sure how technically feasible it is - would be great if it was just a matter of changing the stroke-dasharray but I have a feeling there's more to it then that :)

**Comment by courtneypattison** [comment]
Ugh, filters and transforms are difficult once they start interacting with each other. Unfortunately PDFs are sometimes just raster images so changing lines is difficult.

**Comment by roel-v** [comment]
I've looked into this a bit yesterday; PDF.js has a getSVG() method that would expose the PDF rendered as an SVG, which would let one modify line styles - both changing line styles and thickness and get proper rendering, and things like increase thickness on lines only (not on text). Have you experimented with that before? The current raster-based one would still be in there as a fallback, and there would have to be some way to decide which path to take, I don't have any raster pattern PDFs to look at to check if they have anything other than images.

All pattern PDFs I've seen so far have proper Postscript lines and text, so are easy to convert to browser-native vector representations. But you probably had your reasons to let next-pdf rasterize straight to a canvas?

**Comment by roel-v** [comment]
Also - you may have seen this too from the Facebook post on it, but imagine how next level it would be to integrate https://github.com/Jack000/SVGnest with PP, having some way to select individual paths/parts/pattern pieces and the software then optimally placing it so that you only have to cut shared lines once...

**Comment by roel-v** [comment]
OK never mind just found out that the SVG backend is deprecated and bitrotten already. Sorry for the noise.

**Comment by roel-v** [comment]
Would you be open to modifying the rendering code of PDF.js and using a custom version? In the spirit of 'code or GTFO' I just tried this first before commenting and I have a simple PoC running where I remove all line styling and multiply the line width by 20, which then gives me

![image](https://github.com/Pattern-Projector/pattern-projector/assets/568567/6afaedbe-ef9e-422f-80cb-9e96465aae95)

where the normal rendering of that would be

![image](https://github.com/Pattern-Projector/pattern-projector/assets/568567/7510a593-6372-45f9-933e-ecb283818f5f)

and at the same rendering speed as without the modifications. As you can tell, there's something special going on inside PDF.js to render dashes with different line widths, and it would need some other tweaking, like modifying line width for only some layers, but overall the concept seems to work. But maybe you've tried this also already and there are other issues once you get beyond PoC stage?

Now integrating such a modified library build into the npm build system, next-pdf etc - that is beyond me.

**Comment by hluedeke** [comment]
I have definitely tried this, but did not get as far as you did! I was able to find the render instructions, and I could fetch them as a data structure, but the code to render the structure tree looked pretty involved/hard to reproduce. If you want to take a crack at it, or show us what you did, I think that'd be great! I would definitely prefer the canvas render line weight vs the image manipulation we are currently doing (way more performant).

**Comment by courtneypattison** [comment]
The line thickening has become a lot more performant with Firefox and Chrome with the erode-canvas branch #197, but unfortunately Safari doesn't support canvas.filter using SVG filters and probably never will, so iPad users still have issues with line thickening being done on the CPU. There are also compositing issue with blend-mode on Android #147, so I'm more inclined to go the webgl route for more performant and uniform compositing behavior: unifying the implementation rather than adding polyfills for browsers and devices. That way we can build on top of pdfjs rather than forking the rendering and further split efforts by implementing features for vector or raster PDFs separately. Raster images in patterns is popular with the big 4, and stitching is often required, with no layers, so I think working from rasterized PDFs is the common denominator that keeps things as simple as they can be for the time being. I feel like ere be dragons going the PDF manipulation route, with them being a mix of raster and vector. What do you folks think?

**Comment by roel-v** [comment]
>  or show us what you did,

I will also look into this further, but for the sake of documentation, the way to set up a minimal working example is basically:

```
  mkdir pdfjs_custom
  cd pdfjs_custom
  git clone https://github.com/mozilla/pdf.js.git
  cd pdf.js
  npm install
  npm install --global gulp
  gulp generic
  cd ..
```

Then in that root 'pdfjs_custom' directory, put the three files in the attached zip. Modify the last line of main.js to reference the PDF you want to test on, then launch a localhost server (I used Python's http.server so you'd just do "python server.py" ; other servers can be used, the main thing is to serve the filetype of .mjs files as text/javascript).

This will just show the PDF unmodified. To remove all dashes, you go to pdf.js\src\canvas.js, search for all occurrences of 'setLineDash' and comment them out. Then do 'gulp generic' again, refresh the PDF and that's it.

Then to increase the line width, search for the setLineWidth() function and modify the 'width' variable at the top of that function any way you like. In my screenshots I just multiplied it by 10 or something like that.

[pdfjs_custom.zip](https://github.com/Pattern-Projector/pattern-projector/files/14928962/pdfjs_custom.zip)

**Comment by roel-v** [comment]
> What do you folks think?

I think raster manipulation is always going to be a hack, as you just lose the document semantics once it's rasterized, and there's no way a convolution can make up for that. It's about more than just performance; making dashed lines solid is impossible to do in a generic way on a raster but trivial on the canvas rendering level; how do you deal with antialiased content on the raster? but you don't even have to think about that and it will still work 100% as you'd want it if you make the correct canvas calls in the first place; how do you distinguish between text and lines at the raster level? - also trivial on the canvas level, coloring can be done on the canvas level (although that's less of an issue to do on a raster), ... And that's just for the features that are already in PP.

I don't have any digital big 4 patterns, and none of the patterns I have, have raster images for anything but their logo and things like that. I take your word on what percentage of patterns out there is fundamentally raster based though, you have seen more than I have. But I think it's also a matter of nudging publishers into doing the right thing for pattern sewers. I think Sasha has a document somewhere with hints for publishers on what makes a good projector pattern; maybe there should be a PP wiki page with something similar (I posted some points on Facebook a few days ago when someone asked what the difference is between a 'projector file' and an 'A0' file). Maybe a tool can even be built that 'scores' a pattern on how well-suited it is for use on a projector. I think the future is that publishers will start to take notice of suitability for projecting, and rewarding proper projector-friendly patterns (i.e., having layers, not having raster images, solid line styles, ...) will help that. 'Rewarding' here being - having a superior user experience for users of these patterns. And I do think that modifying the canvas rendering before it gets to the raster stage is a superior user experience in every way.

**Comment by courtneypattison** [comment]
Raster images in PDFs are not ideal, but many patterns are just that. If we could persuade publishers, the preferred format would be pattern semantics and parameters for adjustments i.e. not PDF. I think PP's features should work with the vast number of digital patterns people already have and are available to them today, without them having to seek out projector friendly publishers.

SVG output would be ideal as Safari supports feMorphological filters on SVGs so hopefully that includes images in SVG. It would be good to know why it's deprecated. Failing to revive SVG support, integration points for adjusting line styles and canvas filters on images (with safari pollyfills) in PDFjs would work. Either approach works to avoid deciding if the PDF is vector guides+watermarks with raster patterns or vector patterns with raster logos, since both can be used during rendering. 

This bug has evolved into more of a feature request, and the original issue is resolved in #197 so I'm going to close this issue and create a new one for vector line adjustments in PDF.

**Comment by hluedeke** [comment]
> SVG output would be ideal as Safari supports feMorphological filters on SVGs so hopefully that includes images in SVG. It would be good to know why it's deprecated.

I think native canvas rendering replaced (older) SVG rendering, so they only use it in special cases now.

**Comment by courtneypattison** [comment]
I agree that using a native renderer would be more performant. Although I'm really happy with the performance of Chrome and Firefox since #197! Safari's poor performance issues will become addressable with the switch to webgl we can use for addressing the various other composite issues we have from CSS transforms and blend modes as well. 

It's important to me that we don't lose the current functionality, i.e. line thickening on raster image patterns, and can fix Safari's performance. Projector friendly patterns that follow the requirements set out by the group don't need line thickening or closing, since those are part of the requirements to be considered projector patterns. The patterns that really need these features are the ones with raster images, thin lines, and/or no layers. I cut a pattern without layers last weekend and it was a great experience because I could thicken the lines enough that it was easy to tell which dashed line was my size.

Safari with raster images needs to be well supported for iPad users, and the mix of pollyfills and separate render paths can easily become unmaintainable. If we can create a custom pdfjs renderer to adjust line width, but apply erosions to all images within the pattern with good performance, including Safari, that's great! I think this is the hardest part, since a lot of my PDFs that have raster images and SVG watermarks/lines/text so we wouldn't want to apply the erosions to those parts and thicken them twice.

**Comment by hluedeke** [comment]
> It's important to me that we don't lose the current functionality, i.e. line thickening on raster image patterns, and can fix Safari's performance.

Works for me 👍
